libparsec/crates/testbed/src/template/events.rs

```diff
@@ -1909,96 +1945,323 @@ impl TestbedEventNewShamirRecovery {
         }
     }
 
-    // We need three lifetimes here to describe the fact the output iterator
-    // (lifetime 'c) wraps data from both the self object (lifetime 'a)
-    // and the template (lifetime 'b). Hence 'c outliving both 'a and 'b.
-    pub fn certificates<'a: 'c, 'b: 'c, 'c>(
-        &'a self,
-        template: &'b TestbedTemplate,
-    ) -> impl Iterator<Item = TestbedTemplateEventCertificate> + 'c {
+    fn generate_certificates(
+        &self,
+        template: &TestbedTemplate,
+    ) -> Vec<TestbedTemplateEventCertificate> {
+        let author_signkey = template.device_signing_key(self.author);
+        let author_user_id = template.device_user_id(self.author);
         // One brief certificate + one share certificate per recipient
-        let certifs = self.per_recipient_shares.len() + 1;
+        let certificates_count = self.per_recipient_shares.len() + 1;
+        let mut certificates = Vec::with_capacity(certificates_count);
 
-        (0..certifs).map(move |i| {
-            let mut guard = self.cache.lock().expect("Mutex is poisoned");
+        // Brief certificate
 
-            let populate = || {
-                let author_signkey = template.device_signing_key(self.author);
-                let author_user_id = template.device_user_id(self.author);
-                let mut certifs = Vec::with_capacity(certifs);
+        let certif = ShamirRecoveryBriefCertificate {
+            author: self.author,
+            timestamp: self.timestamp,
+            user_id: author_user_id,
+            threshold: self.threshold,
+            per_recipient_shares: HashMap::from_iter(self.per_recipient_shares.iter().cloned()),
+        };
+        let signed: Bytes = certif.dump_and_sign(author_signkey).into();
 
-                // Brief certificate
+        let brief_certif = TestbedTemplateEventCertificate {
+            certificate: AnyArcCertificate::ShamirRecoveryBrief(Arc::new(certif)),
+            signed_redacted: signed.clone(),
+            signed,
+        };
+        certificates.push(brief_certif);
 
-                let certif = ShamirRecoveryBriefCertificate {
-                    author: self.author,
-                    timestamp: self.timestamp,
-                    user_id: author_user_id,
-                    threshold: self.threshold,
-                    per_recipient_shares: self.per_recipient_shares.clone(),
-                };
-                let signed: Bytes = certif.dump_and_sign(author_signkey).into();
+        // Share certificates
 
-                let brief_certif = TestbedTemplateEventCertificate {
-                    certificate: AnyArcCertificate::ShamirRecoveryBrief(Arc::new(certif)),
-                    signed_redacted: signed.clone(),
-                    signed,
-                };
-                certifs.push(brief_certif);
+        for (recipient, _) in &self.per_recipient_shares {
+            // TODO: Put a real share here once shamir recovery is implemented
```

Comment:
Depends on #8819 or #6090 ?

Reply:
Ah I missed that ! good catch !

Reply:
done.

This snowballed a bit though: I notice some issues with the shamir schemas (incorrect serialization format, inconsistent load doing a signature verification while the dump doesn't sign, no tests) so I also fixed that ;-)